dawn/test/unit-test/dawn/Optimizer/Passes/TestPassRemoveScalars.cpp

```diff
@@ -0,0 +1,587 @@
+//                      / _` |/ _` \ \ /\ / / '_  |
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+#include "dawn/Compiler/DawnCompiler.h"
+#include "dawn/Optimizer/OptimizerContext.h"
+#include "dawn/Optimizer/PassRemoveScalars.h"
+#include "dawn/Unittest/IIRBuilder.h"
+
+#include <gtest/gtest.h>
+
+using namespace dawn;
+
+namespace {
+
+std::shared_ptr<iir::Expr> getRhsOfAssignment(const std::shared_ptr<iir::Stmt> stmt) {
+  if(stmt->getKind() == iir::Stmt::Kind::VarDeclStmt) {
+    const auto& varDeclStmt = std::dynamic_pointer_cast<iir::VarDeclStmt>(stmt);
+    return varDeclStmt->getInitList()[0];
+  } else if(stmt->getKind() == iir::Stmt::Kind::ExprStmt) {
+    const auto& exprStmt = std::dynamic_pointer_cast<iir::ExprStmt>(stmt);
+    if(exprStmt->getExpr()->getKind() == iir::Expr::Kind::AssignmentExpr) {
+      const auto& assignmentExpr =
+          std::dynamic_pointer_cast<iir::AssignmentExpr>(exprStmt->getExpr());
+      return assignmentExpr->getRight();
+    }
+  }
+
+  return nullptr;
+}
+
+iir::DoMethod& getDoMethod(std::shared_ptr<iir::StencilInstantiation>& si) {
+  auto& iir = si->getIIR();
+  auto& stencil = iir->getChild(0);
+  auto& ms = stencil->getChild(0);
+  auto& stage = ms->getChild(0);
+  return *stage->getChild(0);
+}
+
+std::shared_ptr<iir::Stmt> getNthStmt(std::shared_ptr<iir::StencilInstantiation>& si, int n) {
+  return getDoMethod(si).getAST().getStatements()[n];
+}
+
+bool isVarInDoMethodsAccesses(int varAccessID, const iir::DoMethod& doMethod) {
+  for(const auto& stmt : doMethod.getAST().getStatements()) {
+    const auto& access = stmt->getData<iir::IIRStmtData>().CallerAccesses;
+
+    for(auto& accessPair : access->getWriteAccesses()) {
+      if(varAccessID == accessPair.first) { // first = AccessID
+        return true;
+      }
+    }
+    for(const auto& accessPair : access->getReadAccesses()) {
+      if(varAccessID == accessPair.first) { // first = AccessID
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+TEST(TestRemoveScalars, test_unstructured_scalar_01) {
+  using namespace dawn::iir;
+
+  UnstructuredIIRBuilder b;
+  auto f_c = b.field("f_c", ast::LocationType::Cells);
+  auto varA = b.localvar("varA", dawn::BuiltinTypeID::Double, {b.lit(3.0)});
+
+  /// field(cells) f_c;
+  /// double varA = 3.0;
+  /// f_c = varA;
+
+  auto stencil = b.build(
+      "generated",
+      b.stencil(b.multistage(
+          dawn::iir::LoopOrderKind::Forward,
+          b.stage(b.doMethod(dawn::sir::Interval::Start, dawn::sir::Interval::End,
+                             b.declareVar(varA), b.stmt(b.assignExpr(b.at(f_c), b.at(varA))))))));
+
+  // Setup variable's metadata before running pass
```

Comment:
why is this step necessary to be taken manually here?

Reply:
Can extend `IIRBuilder::localvar()` to allow setting the type, so this won't be needed.